src/stan_math_backend/Expression_gen.ml

```diff
@@ -113,6 +113,7 @@ let is_user_dist s =
   ends_with_any user_dist_suffices s
   && not (ends_with_any ["_cdf_log"; "_ccdf_log"] s)
 
+let format_number = String.map ~f:(function '_' -> '\'' | c -> c)
```

Comment:
1) I think it might be slightly better to remove this before we get to the MIR - partial evaluation and related optimizations I think rely on the numbers being valid OCaml numbers
2) This only works in C++14, and we technically need to support C++1y still. I'm not sure if [all of our supported](https://github.com/stan-dev/stan/wiki/Supported-C---Compilers-and-Language-Features) compilers support the `'`.